### DISCUSSION TITLE
Problem with dependencies of dbt_assets

### DISCUSSION THREAD

**Discussion by jordunar** [dagster-io/dagster#18697] [Q&A] [answered]
I'm going to explain my problem: what I am doing right now is migrating to Dagster some processes that were executed before using GCP workflow. Those processes were executed independently (we had one that imports some images, another one that used the result of those imports and did operations with them, etc.) as parts of a bigger pipeline. The thing is that we have a dbt project where the models folder is divided in subfolders, and, for each one of the workflows executed we used different subfolders. For example, for workflow 1 we firstly executed a dbt run command on the subfolder 1 and then launched a dataflow job, and for workflow 2 we firstly executed a dbt run command on the subfolder 2 and then launched two different dataflow jobs. This way, as you can see, we had an amount of steps that were executed sequentially and independently.

What I want now is to merge all this workflows in just one DAG using Dagster. Therefore, what I want to achieve is connecting the last step of workflow 1 (which is launching a dataflow job and can be done with a dagster asset) with the first step of workflow 2 (which is the launching of some dbt models). Anyway, this connection is not a connection that looks for the first step of wf 2 uses the information of the last step of wf1; **it's just a way of indicating Dagster the order of execution of the things**. I have tried to use upstream dependencies (https://docs.dagster.io/integrations/dbt/reference#upstream-dependencies), creating the asset that materializes the last step in workflow 1 (the launch of the dataflow job) as a source and make the first query of the wf2 dbt model to read from it. In the DAG I get what I want (they're propperly connected), but, as you can imagine, the last asset of wf1 is not a table or any source of data (is the launching of a dataflow job) and when I materialize it it fails.

What I want to know if it's possible to do is something like this:
```python

@asset(deps=[launch_test_dataflow], compute_kind="python")
def asset_test1():
    lista = ["a", "b", "c"]
    return Output(
        lista,
        metadata={
            "text_metadata": "longitud lista",
            "lenlista": MetadataValue.json({"len_lista": len(lista)}),
        },
    )


@asset(deps=[asset_test1], compute_kind="python")
def asset_test2():
    lista = ["d", "e", "f"]
    return Output(
        lista,
        metadata={
            "text_metadata": "longitud lista",
            "lenlista": MetadataValue.json({"len_lista": len(lista)}),
        },
    )
```

, but the second asset, instead of being a python asset, being a dbt asset. As you can see the second asset here does not use the information of the first asset; this deps argument just indicates us that it needs to be executed only when the first one is executed.

Thanks in advanced,

Javier

**Comment by rexledesma** [comment]
Try using a sensor. Your sensor can listen for the Dagster job that triggers your Dataflow job. On success, the sensor can launch a job to materialize your dbt assets.

If your Dataflow job is modeled as a Dagster job, you can use a [run status sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#run-status-sensors). If your Dataflow job is modeled as an asset, you can use an [asset sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/asset-sensors#defining-an-asset-sensor).

References:
- [Sensors](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors)

**Comment by jordunar** [comment]
Yes, I had thought about it but I think this is not what I want since I don't want the dbt assets to be launched in a different job; I want them to be launched as assets of the same job but in a different step of the DAG. In the docs it is said that an asset sensor "can be used to kick off a job that computes downstream assets or notifies appropriate stakeholders", but what I want is that, instead of launching a job that computes downstream assets, it still going with the same job but processing the next asset (which would be the dbt assets).

**Reply by rexledesma** [reply]
Currently, the framework cannot do what you are asking for. You want to mix the notion of tasks (triggering the Dataflow job, that is not source of data nor a table) with the notion of assets (your subsets of the dbt project) in the same DAG.

If you don't want to use sensors, then you could just embed the invocation of your dataflow job with your dbt assets. But note that you won't have the dataflow job as part of your DAG, and you won't be able to execute it separately.

```python
@dbt_assets(manifest=dbt_manifest_path, select = 'tag:fook')
def dbt_project_test_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    run_and_wait_for_dataflow_job("my_dataflow_job")
    
    yield from dbt.cli(["build"], context=context).stream()
```

**Reply by jordunar** [reply]
Okey, thank you very much.

**Reply by jordunar** [reply, marked as answer]
Just in case anyone has the same problem I have, I have achieved what I wanted. Instead of modifying the dagster code with any functionality, I have modified the dbt project. I have created a source with the dataflow dagster asset in the models folder (in a yml file) along with the other sources that previously existed as follows :

```
  - name: dagster
    tables:
      - name: import_filter_job__import_wf
        meta:
          dagster:
            asset_key: ["import_filter_job__import_wf"]
```


Then, I also added in the dbt model I wanted to depend on the dataflow job a line specifying that this models depends on that source as follows:

```
-- depends_on: {{ source('dagster','import_filter_job__import_wf') }}

SELECT ...
```

Before specifying it directly on the model I had tried modyfing the dbt_project.yml file to add that dependency on the whole folder that contains this model, but it didn't worked. I selected this concrete model to specify this dependency because I knew it was the one I knew the other models depended on.

**Reply by SeppBerkner** [reply]
I was that anyone :) Thanks for posting this! My use case was that I had a single dlt asset that downloaded data, but due to several reasons I needed to preprocess and split the dataset into two before pushing into my postgres destination. This led to the problem that I only had one dlt asset but two database tables that acted as sources for my dbt models. The connection between the dbt models and the dlt asset was lost and, thus, the data lineage. With your logic by adding dependency on a source that has the name of the dlt asset the pipeline worked again!
